File: src/main/java/td/enastic/emploienastic/model/Matiere.java
package td.enastic.emploienastic.model;

import android.graphics.Color;
import android.os.Build;
import android.util.Log;
import androidx.annotation.RequiresApi;

public class Matiere {

    private String m_name = "";
    private String m_comment="";
    private String m_couleur;
    private int m_id;
    private int m_idEmploi;
    private int m_numeroCase;
    private Teacher m_teacher= new Teacher();
    private String m_niveau;
    private String m_type;
    private String m_teacherName;
    private VColor m_color;

    public Matiere(String name){
        this.m_name = name;
        m_type="CM";
        m_teacherName="";
        m_color = new VColor(255, 255, 255);
        m_numeroCase = m_id = m_idEmploi = -1;
        m_comment = "";
    }

    public String niveau() {
        return m_niveau;
    }

    public void setNiveau(String m_niveau) {
        this.m_niveau = m_niveau;
    }

    public String couleur() {
        return m_couleur;
    }

    public void setCouleur(String m_couleur) {
        this.m_couleur = m_couleur;
    }

    public String teacherName() {
        return m_teacherName;
    }

    public void setTeacherName(String m_teacherName) {
        this.m_teacherName = m_teacherName;
    }

    public Teacher teacher() {
        return m_teacher;
    }

    public void setTeacher(Teacher teacher) {
        this.m_teacher = teacher;
        if(teacher!=null)
        m_teacherName = teacher.getGrade()+" "+teacher.getNom()+"<br/>"+
                teacher.getPrenom();
    }

    public String name() {
        return m_name;
    }
    public void setName(String m_name) {
        this.m_name = m_name;
    }

    public String comment() {
        return m_comment;
    }

    public void setComment(String m_comment) {
        this.m_comment = m_comment;
    }

    public String type() {
        return m_type;
    }

    public void setType(String m_type) {
        this.m_type = m_type;
    }

    public int id() {
        return m_id;
    }

    public void setId(int m_id) {
        this.m_id = m_id;
    }

    public int idemploi() {
        return m_idEmploi;
    }

    public void setIdEmploi(int m_idEmploi) {
        this.m_idEmploi = m_idEmploi;
    }

    public int numeroCase() {
        return m_numeroCase;
    }

    public void setNumeroCase(int m_numeroCase) {
        this.m_numeroCase = m_numeroCase;
    }

    public VColor color() {
        return m_color;
    }

    public void setColor(VColor m_color) {
        this.m_color = m_color;
    }

    public String toString(){
        return m_type+"\n"+
                m_name+ ":"+numeroCase()+"\n"+
                m_teacher.toString()+
                m_color.toString()+"\n"+
                (m_comment.isEmpty() ? Constant.DEFAULT_COMMENT : m_comment)+"\n";
    }

    public static Matiere fromString(String string){
        Matiere m = new Matiere("");
        try {
            String[] tab = string.split("\n");
            int i = 0;
            m.setType(tab[i++]);
            String name = tab[i++];
            m.setTeacher(Teacher.fromString(tab[i++] + "\n" + tab[i++]+ "\n" + tab[i++]));
            try {
                m.setColor(VColor.fromString(tab[i++]));
            } catch (NumberFormatException | IndexOutOfBoundsException e) {
                System.out.println("Matiere.fromString: " + e.getMessage());
            }
            m.setComment((tab[i++].equals(Constant.DEFAULT_COMMENT) ? "" : tab[--i]));
            m.setName(name.split(":")[0]);
            m.setNumeroCase(Integer.parseInt(name.split(":")[1]));
        }catch (IndexOutOfBoundsException e){
            System.out.println("Matiere.fromString: "+e.getMessage());
        }
        return m;
    }

    public String toStyleSheese() {
        return "background-color:" + m_color.toString() + " border: 1px solid black; " +
                "text-align: center; vertical-align: middle;";
    }

    public String toHtml()
    {
        String txt =(m_type.equals("CM") ?"" : m_type) + "<br/>" +
                m_name + "<br/>" +
                m_teacherName+ "<br/>";
        if(m_comment.length()>0)
            txt+= "("+m_comment+")<br/>";
        return "<td style=\" border: 2px solid black; text-align: center; background: "+ m_color.toCssRgbValue()+"\"><div>" + txt + "</div></td>";
    }

    public int getNumeroCase() {
        return m_numeroCase;
    }
};

File: model/matiere.cpp
#include "matiere.h"
#include <QDebug>

Matiere::Matiere(QString name, QObject *parent): QObject(parent), name(name),  typeMatiere("CM"), couleur(QColor(Qt::white)), numeroCase(-1),
    idEmploi(-1), idMatiere(-1), m_commentaire(""){}

Matiere::Matiere(const Matiere &m): QObject(nullptr), name(m.getName()),
   typeMatiere(m.getType()), niveau(m.getNiveau()),
    couleur(m.getCouleur()), numeroCase(m.getNumeroCase()),
    idEmploi(m.getIdEmploi()), idMatiere(m.getIdMatiere()),
    enseignant(m.getEnseignant()), m_commentaire(m.commentaire())
{}

Matiere::~Matiere(){}

const QString &Matiere::getName() const{
    return name;
}

void Matiere::setName(const QString &newName){
    name = newName;
}

const QString &Matiere::getNomEnseigant() const{
    return enseignant.getNom();// + " "+ enseignant.getPrenom();
}

void Matiere::setNomEnseigant(const QString &newNomEnseigant){
    enseignant.setNom(newNomEnseigant);
}

const QColor Matiere::getCouleur() const{
    return couleur;
}

void Matiere::setCouleur(QColor newCouleur){
    couleur = newCouleur;
}

QString Matiere::toString() const{
    if(!name.isEmpty())
        return typeMatiere + " " + name+"\n"+enseignant.toString()+" "+m_commentaire;
    return name+"\n"+enseignant.toString()+" "+m_commentaire;
}

QString Matiere::toStyleSheese()
{
    return QString("QTextEdit{background-color: rgb(%1, %2, %3); border: 1px solid black; "
                    "text-align: center; vertical-align: middle;}").arg(couleur.red()).arg(
                couleur.green()).arg(couleur.blue());
}

QString Matiere::toHtml()
{
    auto txt = "<div style='text-align: center'>" + (name.isEmpty()? "" : (typeMatiere=="CM"?"" : typeMatiere)) + "</div><div style='text-align: center'>" +
            name+"</div><div style='text-align: center'>"+
            enseignant.toString()+"</div>";
    if(!m_commentaire.isEmpty())
        txt+= "<div style='text-align: center'>("+m_commentaire+")</div>";
    return txt;
}

const QString &Matiere::getType() const
{
    return typeMatiere;
}

void Matiere::setType(const QString &newType)
{
    typeMatiere = newType;
}

int Matiere::getNumeroCase() const
{
    return numeroCase;
}

void Matiere::setNumeroCase(int newNumeroCase)
{
    numeroCase = newNumeroCase;
}

const QString &Matiere::getNiveau() const
{
    return niveau;
}

void Matiere::setNiveau(const QString &newNiveau)
{
    niveau = newNiveau;
}

int Matiere::getIdEmploi() const
{
    return idEmploi;
}

void Matiere::setIdEmploi(int newIdEmploi)
{
    idEmploi = newIdEmploi;
}

int Matiere::getIdMatiere() const
{
    return idMatiere;
}

void Matiere::setIdMatiere(int newIdMatiere)
{
    idMatiere = newIdMatiere;
}

const Teacher &Matiere::getEnseignant() const
{
    return enseignant;
}

void Matiere::setEnseignant(const Teacher &newEnseignant)
{
    enseignant = newEnseignant;
}

Matiere& Matiere::operator=(const Matiere &m)
{
    name = m.getName(); typeMatiere = m.getType(); niveau = m.getNiveau();
    couleur = m.getCouleur(); numeroCase = m.getNumeroCase();
    idEmploi = m.getIdEmploi(); idMatiere = m.getIdMatiere();
    enseignant = m.getEnseignant();
    m_commentaire = m.commentaire();
    return *this;
}

const QString Matiere::commentaire() const
{
    return m_commentaire;
}

void Matiere::setCommentaire(const QString &newCommentaire)
{
    m_commentaire = newCommentaire;
}

QDataStream& operator<<(QDataStream &stream, Matiere const& m)
{
    QString comment = m.commentaire().isEmpty() ? DEFAULT_COMMENT : m.commentaire();
    stream<<m.getType()<<m.getName()<<m.getEnseignant()<<m.getCouleur().name()<<comment<<m.getNumeroCase();
    return stream;
}

QDataStream& operator>>(QDataStream &stream, Matiere& m)
{
    QString typeM, name, nomCouleur, comment;
    Teacher enseignant; int numCase=-1;
    stream>>typeM>>name>>enseignant>>nomCouleur>>comment>>numCase;
    m.setType(typeM); m.setName(name); m.setEnseignant(enseignant);
    m.setCouleur(QColor(nomCouleur)); m.setNumeroCase(numCase);
    m.setCommentaire(comment != DEFAULT_COMMENT ? comment : "");
    return stream;
}

QTextStream& operator<<(QTextStream &stream, Matiere const& m)
{
    QString comment = m.commentaire().isEmpty() ? DEFAULT_COMMENT : m.commentaire();
    stream<<m.getType()<<"\n";
    stream<<m.getName()<<"\n";
    stream<<QString("%1").arg(m.getNumeroCase())<<"\n";
    stream<<m.getEnseignant();
    stream<<m.getCouleur().name()<<"\n";
    stream<<comment<<"\n";
    return stream;
}

QTextStream& operator>>(QTextStream &stream, Matiere& m)
{
    QString typeM = stream.readLine(),
            name = stream.readLine(),
            textNumCase = stream.readLine();
    int numCase = textNumCase.toInt();
    Teacher enseignant;
    stream>>enseignant;
    QString nomCouleur = stream.readLine(),
            comment = stream.readLine();
    m.setNumeroCase(numCase);
    m.setType(typeM); m.setName(name); m.setEnseignant(enseignant);
    m.setCouleur(QColor(nomCouleur));
    m.setCommentaire(comment != DEFAULT_COMMENT ? comment : "");
    return stream;
}

bool operator<(Matiere const& m1, Matiere const& m2){
    return m1.getName() < m2.getName();
}
bool operator<=(Matiere const& m1, Matiere const& m2){
    return m1.getName() <= m2.getName();
}
bool operator>(Matiere const& m1, Matiere const& m2){
    return m1.getName() > m2.getName();
}
bool operator>=(Matiere const& m1, Matiere const& m2){
    return m1.getName()>=m2.getName();
}
bool operator==(Matiere const& m1, Matiere const& m2){
    return m1.getName() == m2.getName();
}
